### ISSUE TITLE
C64Studio 7 on Linux using mono

### ISSUE THREAD

**Issue by johanberntsson** [GeorgRottensteiner/C64Studio#56]
I've tried to get C64Studio running on Ubuntu 20.04. There are some copy problems in C64Studio.csproj, but they are easily fixed ny adding
    <PostBuildEvent Condition="'$(OS)' == 'Unix' ">cp $(TargetDir)C64Studio.exe $(SolutionDir)C64StudioRelease/$(TargetFileName)
cp $(TargetDir)*.dll $(SolutionDir)C64StudioRelease
rm -f "$(SolutionDir)C64StudioRelease/BASIC Dialects/*"
cp $(ProjectDir)BASIC?Dialects/* $(SolutionDir)C64StudioRelease/BASIC?Dialects</PostBuildEvent>

I can then build the program. However, when I run it I get

System.TypeInitializationException: The type initializer for 'GR.Image.DPIHandler' threw an exception. ---> System.EntryPointNotFoundException: GetDeviceCaps assembly:<unknown assembly> type:<unknown type> member:(null)
  at (wrapper managed-to-native) GR.Image.DPIHandler.GetDeviceCaps(intptr,int)
  at GR.Image.DPIHandler..cctor () [0x00029] in /home/johan/commodore/C64Studio/Common/Common/DPIHandler.cs:29 
   --- End of inner exception stack trace ---
  at C64Studio.OutputDisplay..ctor () [0x0000d] in /home/johan/commodore/C64Studio/C64Studio/Documents/OutputDisplay.cs:13 
  at (wrapper remoting-invoke-with-check) C64Studio.OutputDisplay..ctor()
  at C64Studio.MainForm..ctor (System.String[] args) [0x00000] in /home/johan/commodore/C64Studio/C64Studio/MainForm.cs:27 
  at (wrapper remoting-invoke-with-check) C64Studio.MainForm..ctor(string[])
  at C64Studio.Program.Main (System.String[] args) [0x00023] in /home/johan/commodore/C64Studio/C64Studio/Program.cs:28 

It seems to be some problem when calling gdi32.dll, which is mapped to libgdiplus.so.0 (see /etc/mono/config), and this library is installed. I don't know much about mono and can't get further, but it would be nice if it could be patched to work on Linux.

**Comment by GeorgRottensteiner** [comment]
Did you install the dotnet35 package? AFAIK this one was required when I tested it via PlayOnLinux in Ubuntu.

**Comment by johanberntsson** [comment]
I'm trying to get it to run on Linux itself, not through wine (which works). The problem seems to be that Xamarin/Mono have dropped support for older versions on .net, and the standard install is only for .net 4 and later. I found some instruction on how to install older versions but they are out of date and don't work any more. I don't know enough to fix this with a reasonable amount of effort.

**Comment by GeorgRottensteiner** [comment]
Ah, good catch. Some people are still using XP, and that's why I postponed updating the used .NET framework (still 3.5)
I think one of these days I have to update anyhow. Ideally I'll switch to .NET Core. That ought to solve this particular problem.

**Comment by johanberntsson** [comment]
That sounds like a good idea. I'll try again when it happens.

**Comment by gcasa** [comment]
Does the same solution apply with macOS?

**Comment by GeorgRottensteiner** [comment]
I have absolutely no idea, but I guess that Mono works the same on different platforms.

**Comment by Beyley** [comment]
when building and running the latest master through latest Mono on Arch Linux, the program builds fine (after changing the post build task), but when running it seems to try to access a folder that only root can, not sure why though

![i](https://i.beyleyisnot.moe/GRHU0Hs.png)

**Comment by Beyley** [comment]
After fixing that problem by creating the folder with the right permissions, i was able to get far enough to see that a lot of code is directly talking with the native Windows APIs, first problem that came up was with the DPIHandler, which i just stubbed out, next was the `FastImage` class, which heavily uses the gdi32.dll native library, it might be possible to gut out the class and replace it with the cross platform `SixLabors.ImageSharp` or to `System.Drawing` though. Overall, most native imports arent *that* important, and it may be possible to replace all of them, as only ~60 native imports are used


> Ah, good catch. Some people are still using XP, and that's why I postponed updating the used .NET framework (still 3.5) I think one of these days I have to update anyhow. Ideally I'll switch to .NET Core. That ought to solve this particular problem.

Switching to .NET Core, would unfortanately not fix the issue, as it still doesnt provide Linux versions of GDI or any other native windows APIs, aside from that, .NET core would also create a bigger issue, that being WinForms, as .NET core has 0 winforms support on linux, so if linux is a target, then you will have to stick with .net framework/mono

**Comment by GeorgRottensteiner** [comment]
Yeah, I did some ugly PInvoke stuff with FastImage. Back when I introduced it I had performance issues and huge delays with System.Drawing. I'll see if I can fix it up now, should be doable.
Imm32.dll is a problem though, that's Windows' input method manager for special keyboard layout methods. As I see it's only used in a few places; I could probably conditionally keep it out when compiling via Mono.

**Comment by Beyley** [comment]
> Yeah, I did some ugly PInvoke stuff with FastImage. Back when I introduced it I had performance issues and huge delays with System.Drawing. I'll see if I can fix it up now, should be doable. Imm32.dll is a problem though, that's Windows' input method manager for special keyboard layout methods. As I see it's only used in a few places; I could probably conditionally keep it out when compiling via Mono.

I believe in Mono, the most popular IMEs do "just work" with the basic Mono WinForms controls, so yeah conditionally skipping the Imm32.dll calls on non-windows platforms would work

**Comment by Beyley** [comment]
Also, if you replace the property group at line 886 in C64Studio.csproj with 
```
  <PropertyGroup>
    <PostBuildEvent Condition="'$(OS)' == 'Windows_NT' ">copy "$(TargetDir)C64Studio.exe" "$(SolutionDir)C64StudioRelease\$(TargetFileName)"
      copy "$(TargetDir)*.dll" "$(SolutionDir)C64StudioRelease"
      del "$(SolutionDir)C64StudioRelease\BASIC Dialects\*.*" /Q
      copy "$(ProjectDir)BASIC Dialects\*.*" "$(SolutionDir)C64StudioRelease\BASIC Dialects"</PostBuildEvent>
    <PostBuildEvent Condition="'$(OS)' == 'Unix' ">cp $(TargetDir)C64Studio.exe $(SolutionDir)C64StudioRelease/$(TargetFileName)
      cp $(TargetDir).dll $(SolutionDir)C64StudioRelease
      rm -f "$(SolutionDir)C64StudioRelease/BASIC Dialects/"
      cp $(ProjectDir)BASIC?Dialects/* $(SolutionDir)C64StudioRelease/BASIC?Dialects</PostBuildEvent>
  </PropertyGroup>
```
ittl fix the builds on linux

**Comment by Beyley** [comment]
hmmm it seems that compile time constant isnt working for me
![i](https://i.beyleyisnot.moe/U7liln4.png)

**Comment by Beyley** [comment]
After gutting out the caret funcs in FastColoredTextBox, i was able to get it fully booting on my local fork which just entirely guts fastimage and related code, so all thats left is to remove FastImage usage, and it should *hopefully* work on linux after that
![i](https://i.beyleyisnot.moe/2uE7ahA.png)

**Comment by GeorgRottensteiner** [comment]
Nice! It'll take a while to replace FastImage. 
Did you check if the cursor is still working in the editor control? The caret is the blinking cursor, that is crucial.

**Comment by Beyley** [comment]
> Nice! It'll take a while to replace FastImage. Did you check if the cursor is still working in the editor control? The caret is the blinking cursor, that is crucial.

![i](https://i.beyleyisnot.moe/n7uW3yw.png)

The position seems wrong, but other then that, it seems to work (albeit it blinks very slowly), i wonder if the position is wrong due to the font?

**Comment by Beyley** [comment]
One problem which is a little more important: the keys always activate the top bar's shortcuts , so letters like `D` are just untypable

**Comment by GeorgRottensteiner** [comment]
The shortcut issue is weird. I wonder if a SetFocus call got missing?

**Comment by Beyley** [comment]
> The shortcut issue is weird. I wonder if a SetFocus call got missing?

Its possible i had accidentally stripped it out when mass deleting code that references FastImage

**Comment by bugjacobs** [comment]
Please keep hacking on this as I really want to use it on Linux with the least bit of fuss :-D
